### PR TITLE
Add support for named parameters in the API

### PR DESCRIPTION
**Pull request by dacort**
Currently, extensions set named parameters directly on the TableFunction object. For rust-based extensions, I'm not sure this is easily doable, so this PR adds two new APIs `duckdb_table_function_add_named_parameter` and `duckdb_bind_get_named_parameter`.

### CODE REVIEW

File: test/api/capi/capi_table_functions.cpp [hunk 1]
```diff
@@ -65,6 +65,7 @@ static void capi_register_table_function(duckdb_connection connection, const cha
 	// add a string parameter
 	duckdb_logical_type type = duckdb_create_logical_type(DUCKDB_TYPE_BIGINT);
 	duckdb_table_function_add_parameter(function, type);
+	duckdb_table_function_add_named_parameter(function, "my_parameter", type);
```
Comment (Mytherin) [inline test/api/capi/capi_table_functions.cpp:68]: Could you add a test where the named parameter is used in a SQL query?

File: src/include/duckdb.h [hunk 2]
```diff
@@ -1710,6 +1710,16 @@ Adds a parameter to the table function.
...
 DUCKDB_API void duckdb_table_function_add_parameter(duckdb_table_function table_function, duckdb_logical_type type);
 
+/*!
+Adds a named parameter to the table function.
+
+* table_function: The table function
+* name: The name of the parameter
+* type: The type of the parameter to add.
+*/
+DUCKDB_API void duckdb_table_function_add_named_parameter(duckdb_table_function table_function, const char *name,
```
Comment (Mytherin) [inline src/include/duckdb.h:1720]: Should there not also be a corresponding `duckdb_bind_get_named_parameter` or similar to allow access to these parameters during the bind phase?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! LGTM. Some comments:

File: test/api/capi/capi_table_functions.cpp (hunk 1 above)
Comment (dacort) [reply test/api/capi/capi_table_functions.cpp:68]: Will do!

File: src/include/duckdb.h (hunk 2 above)
Comment (dacort) [reply src/include/duckdb.h:1720]: Yes, you're right. Have that in a local branch and will push up.

File: test/api/capi/capi_table_functions.cpp (hunk 1 above)
Comment (dacort) [reply test/api/capi/capi_table_functions.cpp]: Updated with an additional test.

File: src/include/duckdb.h (hunk 2 above)
Comment (dacort) [reply src/include/duckdb.h:1720]: Added `duckdb_bind_get_named_parameter`

**Comment by dacort** [comment]
@Mytherin Addressed the comments and all the tests except for the "Python workflow run" succeeded. I'm unsure if that's related - the run [timed out after 6 hours](https://github.com/duckdb/duckdb/actions/runs/4338489074/jobs/7575221063). 

Squashed my changes and running through the test suite again now.

**Comment by Tishj** [comment]
Yea that issue was mentioned, not related to this.
We're still investigating the cause of that problem :)

File: test/api/capi/capi_table_functions.cpp [hunk 3]
```diff
@@ -93,6 +98,10 @@ TEST_CASE("Test Table Functions C API", "[capi]") {
 	REQUIRE_NO_FAIL(*result);
 	REQUIRE(result->Fetch<int64_t>(0, 0) == 42);
 
+	result = tester.Query("SELECT * FROM my_function(1, my_parameter=3)");
```
Comment (Mytherin) [inline test/api/capi/capi_table_functions.cpp:101]: Can we add some tests for the various edge cases as well (e.g. fetching a parameter that isn't there, pushing `nullptr` into the various arguments, etc)?

Comment (dacort) [reply test/api/capi/capi_table_functions.cpp:101]: Sure thing.

Comment (dacort) [reply test/api/capi/capi_table_functions.cpp:101]: @Mytherin Is there any preference for extending the original tests vs. creating a completely new test case? For the additional tests, I would need to update the existing bind/init/function definitions. It shouldn't be a problem, but not sure if we want to just incorporate the new API functionality or leave the existing test cases as is.

Comment (Mytherin) [reply test/api/capi/capi_table_functions.cpp:101]: Perhaps you could create a new test case for this?

Comment (dacort) [reply test/api/capi/capi_table_functions.cpp:101]: Sounds good, thanks for the quick reply!

**Comment by Mytherin** [comment]
Thanks! Looks good to me.

**Pull request merged**